### ISSUE TITLE
Incorrect type error on generics with super bound

### ISSUE THREAD

**Issue by twistedsquare** [typetools/checker-framework#973]
The following Java code, unless I've overlooked something, is valid but gives a general type error when compiled with the checker framework (2.1.5 and 2.1.6):

```
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class SuperConsumer
{
    public boolean next(Consumer<? super List<Object>> consumer)
    {
        List<Object> list = new ArrayList<>();
        consumer.accept(list);
        return true;
    }
}
```

I get:
```
Error: java: [argument.type.incompatible] incompatible types in argument.
  found   : List<Object>
  required: ? extends Object super List<Object>
```

Thankfully, the larger example in my program is an overriding method, and I was able to use skipDefs to ignore that package and its descendants.  But as you can see above, it appears even without any non-standard libraries or any annotations.

**Comment by mernst** [comment]
Running the checker with the `-AprintAllQualifiers` command-line option shows the real problem:

```
SuperConsumer.java:10: error: [argument.type.incompatible] incompatible types in argument.
        consumer.accept(list);
                        ^
  found   : @UnknownKeyFor List<@UnknownKeyFor Object>
  required: ? extends @UnknownKeyFor Object super @KeyForBottom List<@KeyForBottom Object>
1 error
```

(I'll open a different issue about the fact that in the error message, two occurrences of `List<Object>` were printed identically even though their types differed.)

**Comment by smillst** [comment]
This error message is expected, though as Mike pointed, out confusing. It is also complicated by a bug where generics types in wildcard bounds are not checked (Issue #849). So, let's look a similar example that doesn't have a generic wildcard bound:
````
import java.util.function.Consumer;

public class SuperConsumer {
    public boolean next2(Consumer<? super String> consumer) {
        consumer.accept("");
        return true;
    }
}  
````
produces this following error (when using -AprintAllQualifiers):
````
SuperConsumer.java:8: error: [argument.type.incompatible] incompatible types in argument.
        consumer.accept("");
                        ^
  found   : @UnknownKeyFor String
  required: ? extends @UnknownKeyFor Object super @KeyForBottom String
````
The problem is that the argument to accept must be a subtype of `@KeyForBottom String`, but it isn't.  One way to fix this warning is add `@UnknownKeyFor` to the lower bound of the wildcard: `Consumer<? super @KeyForUnknown String>`.   You could also change the default for explicit lower bounds using a [`@DefaultQualifier`](http://types.cs.washington.edu/checker-framework/current/api/org/checkerframework/framework/qual/DefaultQualifier.html)`(value = UnknownKeyFor.class, locations = TypeUseLocation.EXPLICIT_LOWER_BOUND)` annotation on the package which you skipped.

**Comment by twistedsquare** [comment]
Thanks for your response.  I now understand what the error is -- apologies for the false alarm, but as you can understand, I was baffled by the original message!  I will adjust my code accordingly.
